pkg/cortex/cortex.go

```diff
@@ -193,6 +199,33 @@ func (c *Config) Validate(log log.Logger) error {
 	return nil
 }
 
+// validateYAMLEmptyNodes ensure that no empty node has been specified in the YAML config file.
+// When an empty node is defined in YAML, the YAML parser sets the whole struct to its zero value
+// and so we loose all default values. It's very difficult to detect this case for the user, so we
+// try to prevent it (on the root level) with this custom validation.
+func (c *Config) validateYAMLEmptyNodes() error {
+	defaults := Config{}
+	flagext.DefaultValues(&defaults)
+
+	defStruct := reflect.ValueOf(defaults)
+	cfgStruct := reflect.ValueOf(*c)
+
+	// We expect all structs are the exact same. This check should never fail.
+	if cfgStruct.NumField() != defStruct.NumField() {
+		return errors.New("unable to validate configuration because of mismatching internal config data structure")
+	}
+
+	for i := 0; i < cfgStruct.NumField(); i++ {
```

Comment:
I expected to see recursive checks for nested structs. Can this problem happen with nested structs as well?

Reply:
This problem can happen with any struct deserialised from JSON. The problem is that the deeper we go, the more likely we hit the case "all zero values" is a valid config. After a which look at the configs at root level, I believe this is not a problem.